src/FreeAge/common/type_stats.hpp

```diff
@@ -0,0 +1,167 @@
+// Copyright 2020 The FreeAge authors
+// This file is part of FreeAge, licensed under the new BSD license.
+// See the COPYING file in the project root for the license text.
+
+#pragma once
+
+#include <QRect>
+#include <QSize>
+
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/resources.hpp"
+
+static constexpr u32 typeStatsVersion = 1;
+
+enum class AttackType {
+  NoAttack = 0,
+  /// TODO: rename ?
+  Default,
+  /// Full damage to all affected in the area of effect.
+  HomogeneousAreaOfEffect,
+  /// Full damage to primary target and 0.5 to all others in the area of effect.
+  TrampleHalf, // War Elephants, Battle Elephant
+  /// Full damage to primary target and 5 unaffected by armor damage in the area of effect.
+  TrampleFivePure, // Cataphracts, Druzhina
+  /// First projectile does full damage and other projectiles deal 3 pierce damage and 0 melee damage.
+  MultipleProjectiles, // Chu Ko Nu, Kipchak
+  
+  TownCenter, Castle, Tower, BombardTower
+};
+
+enum class GarrisonType {
+  /// No unit can be garrison.
+  NoGarrison = 0,
+  /// All units.
+  AllUnits, // Transport ships
+  /// Villagers, monks and foot soldiers only.
+  VillagersAndMonksAndFootSoldiers, // TownCenter and towers
+  /// Villagers and foot soldiers only.
+  VillagersFootSoldiers, // Rams and Siege Towers
+  /// All units except siege units.
+  NonSiege, // Castle
+  /// Villagers only.
+  Villagers, // Khmer houses
+  /// Units from production buildings can be trained directly into garrison,
+  /// but not garrison from outside.
+  Production,
+  /// Same as Production plus Relics.
+  ProductionAndRelics
+};
+
+struct PopulationCount {
+  /// TODO (puzzlepaint): Refactor ?
+  
+  inline PopulationCount()
+    : doubledCount(0) {};
+  
+  inline void SetToIntegerPopulationSpace(int count) { doubledCount = count * 2; }
+  inline void SetToIntegerPopulationDemand(int count) { doubledCount = -count * 2; }
+  inline void SetToOneHalfPopulationDemand() { doubledCount = -1; }
+
+  inline int GetDoubledPopulationSpace() const { return doubledCount > 0 ? doubledCount : 0; }
+  inline int GetDoubledPopulationDemand() const { return doubledCount < 0 ? -doubledCount : 0; }
+
+ //private:
+
+  /// The doubled population space if positive or demand if negative.
+  int doubledCount;
+};
+
+struct ObjectTypeStats {
```

Comment:
Maybe it could make sense to add `// TODO: not used by the game yet` to each of the attributes that is not actually used by the game yet to keep an overview of that